### PR TITLE
Feature/acceptance tests

### PR DESCRIPTION
**Pull request by mike927**
Acceptance tests for Scanner class

### CODE REVIEW

File: local_file_spec.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
+require_relative 'spec_helper'
+require_relative 'local_file'
+
+describe LocalFile do
+  let(:file1) { LocalFile.new(Tempfile.new('temp1')) }
+  let(:file2) { LocalFile.new(Tempfile.new('temp2')) }
+  let(:output) { 50 }
```
Comment (spidzio) [inline local_file_spec.rb]: You are using `output` variable only for `#size_dir` spec. Define it there in `describe '#size_dir'` block.

File: local_file_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  let(:threshold) { 60 }
```
Comment (spidzio) [inline local_file_spec.rb]: The same as above for `threshold` - used only for `#size_diff` spec.

File: scanner.rb [hunk 3]
```diff
@@ -3,37 +3,45 @@
...
-    print_similar
+    create_file_set(@dirname)
+    puts print_similar(@files)
   end
 
-  private
-
   def check_dir(dirname)
-    fail 'Wrong directory' unless File.exist?(dirname) || File.directory?(dirname)
+    # fail 'Wrong directory' unless File.directory?(dirname) && File.exist?(dirname)
```
Comment (spidzio) [inline scanner.rb]: Never leave commented parts of code, please.
If someone needs to check previous versions - it is possible via git history.

File: scanner.rb [hunk 4]
```diff
@@ -3,37 +3,45 @@
...
+  def create_file_set(dirname)
+    Dir.glob("#{dirname}/**/*") do |file|
       @files << LocalFile.new(file)
     end
+    @files
   end
 
-  def print_similar
-    @files.combination(2).each do |pair|
+  def print_similar(files)
```
Comment (spidzio) [inline scanner.rb]: This method doesn't print anything.
If you have changed it for tests purposes, you have 2 options: change the method name, or restore previous code and use `output.to_stdout` matcher in specs.

File: scanner_spec.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+describe Scanner do
+  let(:name) { 'somedir' }
+  let(:basic_path) { "#{Dir.pwd}/#{name}" }
+  let(:dir) { Dir.new(name) }
+  let(:path) { dir.path }
+  let(:create_newfiles) do
+    (1..3).each do |i|
+      File.new('somedir/temp' + i.to_s + '.txt', 'w+')
+    end
+  end
```
Comment (spidzio) [inline scanner_spec.rb]: You should not use `let` to prepare environment for tests. 
There is `before` and `after` for such cases.

File: scanner_spec.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  let(:set_of_localfiles) do
+    file.create_file_set(name)
+  end
```
Comment (spidzio) [inline scanner_spec.rb]: Same as above

File: scanner_spec.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  let(:file) { Scanner.new(dirname: path, size_treshold: 10) }
```
Comment (spidzio) [inline scanner_spec.rb]: This is subject of spec - please define it as subject.

File: scanner_spec.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+     basic_path + '/temp2.txt <---> ' + basic_path + '/temp1.txt',
+     basic_path + '/temp3.txt <---> ' + basic_path + '/temp1.txt']
+  end
+  let(:remove) do
+    Dir.foreach(name) do |file|
+      path = File.join(name, file)
+      File.delete(path) if file != '.' && file != '..'
+    end
+    Dir.delete(name)
+  end
```
Comment (spidzio) [inline scanner_spec.rb]: You should use `after` command to clean the testing environment after tests.

File: scanner_spec.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  before(:each) { Dir.mkdir(name) unless File.exist?(name) }
```
Comment (spidzio) [inline scanner_spec.rb]: No need to `before(:each)`, just `before`

File: scanner_spec.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  after(:each) do
```
Comment (spidzio) [inline scanner_spec.rb]: the same as above

File: local_file_spec.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        file2.name = 'other_sample'
+        file1.size = 20
+        file2.size = 70
+        file1.ext = '.txt'
+        file2.ext = '.rb'
+        expect(file1.similar_to?(file2, threshold)).to be false
+      end
+    end
+  end
+end
```
Comment (spidzio) [inline local_file_spec.rb]: This spec file is really good. Only 2 simple styling fixes. 
Good usage of contexts.

**Comment by mike927** [comment]
I've made changes, if everything seems to be all right I will merge pull request

File: scanner.rb [hunk 12]
```diff
@@ -3,37 +3,44 @@
...
     @size_treshold = size_treshold.to_i
     @files = []
+    @duplicates  = []
   end
 
   def run
-    create_file_set
-    print_similar
+    create_file_set(@dirname)
+    puts find_similar(@files)
```
Comment (spidzio) [inline scanner.rb]: You don't need to pass arguments since they are attributes.
Use attributes in methods.

File: scanner.rb [hunk 13]
```diff
@@ -3,37 +3,44 @@
...
+    if File.directory?(dirname) && File.exist?(dirname)
+      return true
+    else
+      fail 'Wrong directory'
+    end
   end
 
-  def create_file_set
-    Dir.glob("#{@dirname}/**/*") do |file|
+  def create_file_set(dirname)
```
Comment (spidzio) [inline scanner.rb]: use attribute not parameter

File: scanner.rb [hunk 14]
```diff
@@ -3,37 +3,44 @@
...
+    Dir.glob("#{dirname}/**/*") do |file|
       @files << LocalFile.new(file)
     end
+    @files
   end
 
-  def print_similar
-    @files.combination(2).each do |pair|
+  def find_similar(files)
```
Comment (spidzio) [inline scanner.rb]: use attribute not parameter

File: scanner_spec.rb [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+      path = File.join(name, file)
+      File.delete(path) if file != '.' && file != '..'
+    end
+    Dir.delete(name)
+  end
+
+  describe '#check_dir' do
+    it do
+      expect(scanner.check_dir(path)).to be true
+    end
```
Comment (spidzio) [inline scanner_spec.rb]:
```ruby
it { expect(scanner.check_dir(path)).to be true }
```

File: scanner_spec.rb [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  describe '#create_file_set' do
+    before { create_tmp_files }
+    after { clean_tmp_files }
+    it do
+      scanner.create_file_set(name).each do |file|
+        expect(file).to be_a LocalFile
+      end
```
Comment (spidzio) [inline scanner_spec.rb]: You can also check it file set has proper size.

File: scanner_spec.rb [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+  let(:name) { 'somedir' }
+  let(:basic_path) { "#{Dir.pwd}/#{name}" }
+  let(:dir) { Dir.new(name) }
+  let(:path) { dir.path }
+  let(:set_of_localfiles) { scanner.create_file_set(name) }
+  let(:output) do
+    [basic_path + '/temp2.txt <---> ' + basic_path + '/temp3.txt',
+     basic_path + '/temp2.txt <---> ' + basic_path + '/temp1.txt',
+     basic_path + '/temp3.txt <---> ' + basic_path + '/temp1.txt']
+  end
```
Comment (spidzio) [inline scanner_spec.rb]: You are using this only in `describe '#find_similar' do`
Move it there.

File: scanner.rb [hunk 18]
```diff
@@ -3,37 +3,44 @@
...
+    files.combination(2).each do |pair|
       f1, f2 = pair
-      puts "#{f1.path} <---> #{f2.path}" if f1.similar_to?(f2, @size_treshold)
+      @duplicates << "#{f1.path} <---> #{f2.path}" if f1.similar_to?(f2, @size_treshold)
```
Comment (spidzio) [inline scanner.rb]: I don't like this solution. You are not printing it now, but making some formatted data which is environment dependency, and shouldn't have happen. If I would like to use this class, and output the data not to stdout, but somewhere else - I would need to parse similarities you are giving me.

The method should be called `find_similarities` and it should return array of pairs.

File: scanner.rb (hunk 14 above)
Comment (mike927) [inline scanner.rb]: I use arguments instead of attributes becuse it was easier to test methods which take arguments than with any

Comment (spidzio) [inline scanner.rb]: Yes,  but it is not correct way. Tests should work for us, not the other way.

**Comment by mike927** [comment]
I've added further fixes

**Comment by spidzio** [comment]
:+1:

**Pull request merged**